moveit_ros/planning_interface/test/test_serialize_msg_python.cpp

```diff
@@ -0,0 +1,371 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, RWTH Aachen University.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of RWTH Aachen University nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Bjarne von Horn */
+
+#include <Python.h>
+#include <boost/python.hpp>
+#include <gtest/gtest.h>
+#include <moveit/py_bindings_tools/serialize_msg.h>
+#include <geometry_msgs/Vector3.h>
+#include <cstring>
+
+namespace bp = boost::python;
+using moveit::py_bindings_tools::ByteString;
+
+struct ByteStringTestHelper
+{
+  bool do_compare(const std::vector<unsigned char>& data, PyObject* obj)
+  {
+    const char* py_data = PyBytes_AsString(obj);
+    if (!py_data)
+      return false;
+    Py_ssize_t size = PyBytes_GET_SIZE(obj);
+    if (size != data.size())
+      return false;
+    return std::memcmp(py_data, &data[0], size) == 0;
+  }
+  bool compare_bytes(const ByteString& s)
+  {
+    const std::vector<unsigned char> testdata{ 0xff, 0xef, 0x00, 0x10 };
```

Comment:
What is actually test here? An arbitrary? Bytestring against a fixed uchar vector?

Reply:
this method is exposed to Python and called with `b'\\xff\\xef\\x00\\x10'` as argument from within Python. Then the `bytes` Python instance will be converted to C++ `py_bindings_tools::ByteString` by boost, the converted Object is then tested.
So we know that embedded NULL's won't be dropped.